$OpenBSD: patch-tools_clang_lib_Basic_Targets_cpp,v 1.21 2017/02/20 07:04:00 ajacoutot Exp $

- Enable support for __float128 in Clang on OpenBSD/X86
- Set ABI for ARM
- [OpenBSD] set OpenBSD defines for aarch64

--- tools/clang/lib/Basic/Targets.cpp.orig	Wed Feb  8 20:35:17 2017
+++ tools/clang/lib/Basic/Targets.cpp	Sun Feb 19 17:36:20 2017
@@ -545,6 +545,8 @@ class OpenBSDTargetInfo : public OSTargetInfo<Target> 
     Builder.defineMacro("__ELF__");
     if (Opts.POSIXThreads)
       Builder.defineMacro("_REENTRANT");
+    if (this->HasFloat128)
+      Builder.defineMacro("__FLOAT128__");
   }
 public:
   OpenBSDTargetInfo(const llvm::Triple &Triple, const TargetOptions &Opts)
@@ -552,11 +554,11 @@ class OpenBSDTargetInfo : public OSTargetInfo<Target> 
     this->TLSSupported = false;
 
       switch (Triple.getArch()) {
-        default:
         case llvm::Triple::x86:
         case llvm::Triple::x86_64:
-        case llvm::Triple::arm:
-        case llvm::Triple::sparc:
+          this->HasFloat128 = true;
+          // FALLTHROUGH
+        default:
           this->MCountName = "__mcount";
           break;
         case llvm::Triple::mips64:
@@ -5144,6 +5146,8 @@ class ARMTargetInfo : public TargetInfo { (public)
       default:
         if (Triple.getOS() == llvm::Triple::NetBSD)
           setABI("apcs-gnu");
+        else if (Triple.getOS() == llvm::Triple::OpenBSD)
+          setABI("aapcs-linux");
         else
           setABI("aapcs");
         break;
@@ -8574,6 +8578,8 @@ static TargetInfo *AllocateTarget(const llvm::Triple &
       return new LinuxTargetInfo<AArch64leTargetInfo>(Triple, Opts);
     case llvm::Triple::NetBSD:
       return new NetBSDTargetInfo<AArch64leTargetInfo>(Triple, Opts);
+    case llvm::Triple::OpenBSD:
+      return new OpenBSDTargetInfo<AArch64leTargetInfo>(Triple, Opts);
     default:
       return new AArch64leTargetInfo(Triple, Opts);
     }
